edge-modules/TestAnalyzer/Reporter.cs

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace TestAnalyzer
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using Microsoft.Azure.Devices.Edge.ModuleUtil;
+    using Microsoft.Extensions.Logging;
+
+    class Reporter
+    {
+        static readonly ILogger Logger = ModuleUtil.CreateLogger("Analyzer");
+
+        public static DeviceAnalysis GetDeviceReport(double toleranceInMilliseconds)
+        {
+            return new DeviceAnalysis(GetReceivedMessagesReport(toleranceInMilliseconds), GetDirectMethodsReport(), GetTwinsReport());
+        }
+
+        static IList<ResponseOrientedReport> GetDirectMethodsReport()
+        {
+            IDictionary<string, IDictionary<string, Tuple<int, DateTime>>> dms = ReportingCache.Instance.GetDirectMethodsSnapshot();
+            string description = "Report for direct methods";
+            return GetReportHelper(dms, description);
+        }
+
+        static IList<ResponseOrientedReport> GetTwinsReport()
+        {
+            IDictionary<string, IDictionary<string, Tuple<int, DateTime>>> twins = ReportingCache.Instance.GetTwinsSnapshot();
+            string description = "Report for twins";
+            return GetReportHelper(twins, description);
+        }
+
+        static IList<ResponseOrientedReport> GetReportHelper(IDictionary<string, IDictionary<string, Tuple<int, DateTime>>> cache, string reportDescription)
```

Comment:
what is cache? why name this method as "GetReportHelper" but it doesn't return any report helper instance.

Reply:
Cache is the [data structure](https://github.com/and-rewsmith/iotedge/blob/andsmi/longhaul-twin-dm/edge-modules/TestAnalyzer/ReportingCache.cs#L14) we are using to store the test reports. GetReportHelper() is a helper function that processes the shared operations.